src/samplers/disk/mod.rs

```diff
@@ -139,36 +141,51 @@ impl Disk {
                     include_str!("../../common/value_to_index2.c"),
                 );
                 let mut bpf = bcc::BPF::new(&code)?;
-                // load + attach kprobes!
-                bcc::Kprobe::new()
-                    .handler("trace_pid_start")
-                    .function("blk_account_io_start")
-                    .attach(&mut bpf)?;
-                if let Ok(results) = bpf.get_kprobe_functions("blk_start_request") {
-                    if !results.is_empty() {
-                        bcc::Kprobe::new()
-                            .handler("trace_req_start")
-                            .function("blk_start_request")
-                            .attach(&mut bpf)?;
+                // collect the set of probes required from the statistics enabled.
+                let mut probes = HashSet::new();
+                for statistic in &self.statistics {
+                    for probe in statistic.bpf_probes_required() {
+                        probes.insert(probe);
                     }
                 }
-                bcc::Kprobe::new()
-                    .handler("trace_req_start")
-                    .function("blk_mq_start_request")
-                    .attach(&mut bpf)?;
-                if let Ok(results) = bpf.get_kprobe_functions("blk_account_io_completion") {
-                    if !results.is_empty() {
-                        bcc::Kprobe::new()
-                            .handler("do_count")
-                            .function("blk_account_io_completion")
-                            .attach(&mut bpf)?;
-                    } else {
-                        bcc::Kprobe::new()
-                            .handler("do_count")
-                            .function("blk_account_io_done")
-                            .attach(&mut bpf)?;
+
+                // load + attach the kernel probes that are required to the bpf instance.
+                for probe in probes {
+                    match probe.name.as_str() {
+                        "blk_start_request" => {
+                            // respect the current logic, attach only if 'blk_start_request' can be found.
+                            if let Ok(results) = bpf.get_kprobe_functions("blk_start_request") {
+                                if !results.is_empty() {
+                                    probe.try_attach_to_bpf(&mut bpf)?
+                                }
+                            }
+                        }
+                        "blk_account_io_completion" =>
+                        // respect the current logic, if 'blk_account_io_completion' exists, we attach this probe.
```

Comment:
Just a note, but once we merge this change this will become the new current logic.
We should probably remove the `respect the current logic` part.

Reply:
yeah, what I am trying to do here is to still follow what the old code is doing.
Maybe  I should remove the 'respect the current logic' comment.